### PR TITLE
Remove cluster_id from cluster node's hostname

### PR DESCRIPTION
**Pull request by cs-zhang**
Fixed: #308 

Signed-off-by: CS Zhang <zhangcho@us.ibm.com>

### CODE REVIEW

**Comment by bpradipt** [comment]
/hold
We'll need to address the following requirements and simplify the handling of the names

1. OCP node names having region details
2. PowerVS resources - instance, boot volumes, network ids, keys etc having user specified string (eg user shortid). 
3. No portion of the name should be repeated in FQDN 
4. PowerVS resource names can't contain dot (.)

Example
cluster_domain: mydom.com
node_name: bootstrap, masterN, workerN (eventually we should allow any names as the node role indicate the type of node)
cluster_id: test  (sub-domain)   

Example fqdn for bootstrap: bootstrap.test.mydom.com  <resource_name>[.<cluster_id>].<cluster_domain>

where,
 `resource_name == node_name`

Now if we augment the `<resource_name>` and `cluster_id` to handle requirement 1 and 2, then

resource_name = [<node_prefix>-]<node_name>[-<cluster_id_prefix>]
cluster_id_prefix = [<region>]-[<cluster_id_prefix>]
cluster_id = [<cluster_id_prefix>.]<cluster_id>

`[ ] indicates optional param`

where,
node_prefix = `<user_specified_node_prefix>`  or null
cluster_id_prefix = `<random-hex>` or null


Example-1

node_prefix: bpradipt
region: eu-de-2
cluster_id_prefix: false
cluster_id: test
cluster_domain: mydom.com

master resource name: bpradipt-master0-eu-de-2
master fqdn:  bpradipt-master0-eu-de-2.test.mydom.com
OCP node name: bpradipt-master0-eu-de-2.test


Example-2
node_prefix: bpradipt-dev
region: eu-de-2
cluster_id_prefix: true
cluster_id: test
cluster_domain: mydom.com

master resource name: bpradipt-dev-master0-eu-de-2-idk4
master fqdn:  bpradipt-dev-master0-eu-de-2-idk4.test.mydom.com
OCP node name: bpradipt-dev-master0-eu-de-2-idk4.test



cc @cs-zhang @Prajyot-Parab @yussufsh please review and let me know your comments

**Comment by bpradipt** [comment]
@cs-zhang we can go with example-3. Let's wait for a day to hear back from others.

**Comment by cs-zhang** [comment]
For resource name, all resources belong to same cluster should be grouped together showing in UI.  For `Example 3`: `master0-bpradipt-dev-eu-de-2-idk4`  that would not be possible, in this case the resources will be grouped by base node name `master0`. So I prefer `bpradipt-dev-eu-de-2-idk4-master0`, that will make all resources will be grouped together.

Example 1 -- defined cluster_id 
node_prefix: bpradipt-dev
region: eu-de-2
cluster_id_prefix: true
cluster_id: test
cluster_domain: mydom.com

master resource name: bpradipt-dev-eu-de-2-idk4-master0
master fqdn: bpradipt-dev-eu-de-2-idk4-master0.test.mydom.com
OCP node name: bpradipt-dev-eu-de-2-idk4-master0.test

Example 2 -- undefined cluster_id
node_prefix: bpradipt-dev
region: eu-de-2
cluster_id_prefix: cluster-id
cluster_id: false
cluster_domain: mydom.com

master resource name: bpradipt-dev-eu-de-2-idk4-master0
master fqdn: bpradipt-dev-eu-de-2-idk4-master0.cluster-id-xxxx.mydom.com
OCP node name: bpradipt-dev-eu-de-2-idk4-master0.cluster-id-xxxx

**Comment by bpradipt** [comment]
> For resource name, all resources belong to same cluster should be grouped together showing in UI. For `Example 3`: `master0-bpradipt-dev-eu-de-2-idk4` that would not be possible, in this case the resources will be grouped by base node name `master0`. So I prefer `bpradipt-dev-eu-de-2-idk4-master0`, that will make all resources will be grouped together.
> 
Good point @cs-zhang .

> Example 1 -- defined cluster_id
> node_prefix: bpradipt-dev
> region: eu-de-2
> cluster_id_prefix: true
> cluster_id: test
> cluster_domain: mydom.com
> 
> master resource name: bpradipt-dev-eu-de-2-idk4-master0
> master fqdn: bpradipt-dev-eu-de-2-idk4-master0.test.mydom.com
> OCP node name: bpradipt-dev-eu-de-2-idk4-master0.test
> 
> Example 2 -- undefined cluster_id
> node_prefix: bpradipt-dev
> region: eu-de-2
> cluster_id_prefix: cluster-id

I think you meant it to be true `cluster_id_prefix: true`

> cluster_id: false

undefined cluster_id should be empty `cluster_id: `

> cluster_domain: mydom.com
> 
> master resource name: bpradipt-dev-eu-de-2-idk4-master0

> master fqdn: bpradipt-dev-eu-de-2-idk4-master0.cluster-id-xxxx.mydom.com

this should be `bpradipt-dev-eu-de-2-idk4-master0.mydom.com`

> OCP node name: bpradipt-dev-eu-de-2-idk4-master0.cluster-id-xxxx

this should be `bpradipt-dev-eu-de-2-idk4-master0`

**Comment by yussufsh** [comment]
I feel we can still use the old format for hostnames eg. `master-0.<cluster_id>-<zone>.<domain>` and set full FQDN in the dhcp configuration in the [helpernode vars](https://github.com/ocp-power-automation/ocp4-upi-powervs/blob/master/modules/5_install/install.tf#L60).

`oc get nodes` should start showing node names with sub and main domain. Can you try it our @cs-zhang ?

**Comment by bpradipt** [comment]
> `cluster_id_prefix: true` made me confused all along and there is no `node_prefix`.
> 
> `cluster_id_prefix` can take a string value or can be empty string. 

This is fine as long as we don't use a random hex string when `cluster_id_prefix` is an empty string. The reason I mentioned it as boolean was to emphasize the point about controlling the inclusion (or exclusion) of hex string.

**Comment by yussufsh** [comment]
> This is fine as long as we don't use a random hex string when `cluster_id_prefix` is an empty string. The reason I mentioned it as boolean was to emphasize the point about controlling the inclusion (or exclusion) of hex string.

The hex is created when cluster_id is not provided. If cluster_id_prefix is empty only cluster_id is used (no hex). Inclusion of zone depends on the another flag.

**Comment by cs-zhang** [comment]
Now the instance name in PowerVS looks like this: `cz-d8e4-eu-de-2-master-0`
the FQDN is used as node name, and the output of the `oc`:
```
[root@cz-d8e4-eu-de-2-bastion-0 ~]# oc get nodes
NAME                                 STATUS   ROLES    AGE    VERSION
eu-de-2-master-0.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
eu-de-2-master-1.cz-d8e4.cloud.lab   Ready    master   111m   v1.20.0+7d0a2b2
eu-de-2-master-2.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
eu-de-2-worker-0.cz-d8e4.cloud.lab   Ready    worker   96m    v1.20.0+7d0a2b2
eu-de-2-worker-1.cz-d8e4.cloud.lab   Ready    worker   95m    v1.20.0+7d0a2b2
[root@cz-d8e4-eu-de-2-bastion-0 ~]#
```

**Comment by yussufsh** [comment]
> Now the instance name in PowerVS looks like this: `cz-d8e4-eu-de-2-master-0`
> the FQDN is used as node name, and the output of the `oc`:
> 
> ```
> [root@cz-d8e4-eu-de-2-bastion-0 ~]# oc get nodes
> NAME                                 STATUS   ROLES    AGE    VERSION
> eu-de-2-master-0.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
> eu-de-2-master-1.cz-d8e4.cloud.lab   Ready    master   111m   v1.20.0+7d0a2b2
> eu-de-2-master-2.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
> eu-de-2-worker-0.cz-d8e4.cloud.lab   Ready    worker   96m    v1.20.0+7d0a2b2
> eu-de-2-worker-1.cz-d8e4.cloud.lab   Ready    worker   95m    v1.20.0+7d0a2b2
> [root@cz-d8e4-eu-de-2-bastion-0 ~]#
> ```

I think this fits all the use cases. @bpradipt are we okay with it? 

Suggestion:
`eu-de-2-master-0.cz-d8e4.cloud.lab` -> `master-0.cz-d8e4-eu-de-2.cloud.lab`
The region info can be added to the full cluster name and will be easy to implement.

**Comment by bpradipt** [comment]
> > Now the instance name in PowerVS looks like this: `cz-d8e4-eu-de-2-master-0`
> > the FQDN is used as node name, and the output of the `oc`:
> > ```
> > [root@cz-d8e4-eu-de-2-bastion-0 ~]# oc get nodes
> > NAME                                 STATUS   ROLES    AGE    VERSION
> > eu-de-2-master-0.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
> > eu-de-2-master-1.cz-d8e4.cloud.lab   Ready    master   111m   v1.20.0+7d0a2b2
> > eu-de-2-master-2.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
> > eu-de-2-worker-0.cz-d8e4.cloud.lab   Ready    worker   96m    v1.20.0+7d0a2b2
> > eu-de-2-worker-1.cz-d8e4.cloud.lab   Ready    worker   95m    v1.20.0+7d0a2b2
> > [root@cz-d8e4-eu-de-2-bastion-0 ~]#
> > ```

Can cluster-id be empty ? For the case where user doesn't want the hexID

> 
> I think this fits all the use cases. @bpradipt are we okay with it?
> 
> Suggestion:
> `eu-de-2-master-0.cz-d8e4.cloud.lab` -> `master-0.cz-d8e4-eu-de-2.cloud.lab`
> The region info can be added to the full cluster name and will be easy to implement.

For the above, @cs-zhang mentioned the following "For resource name, all resources belong to same cluster should be grouped together showing in UI. For Example 3: master0-bpradipt-dev-eu-de-2-idk4 that would not be possible, in this case the resources will be grouped by base node name master0. So I prefer bpradipt-dev-eu-de-2-idk4-master0, that will make all resources will be grouped together."

**Comment by yussufsh** [comment]
> 
> 
> > > Now the instance name in PowerVS looks like this: `cz-d8e4-eu-de-2-master-0`
> > > the FQDN is used as node name, and the output of the `oc`:
> > > ```
> > > [root@cz-d8e4-eu-de-2-bastion-0 ~]# oc get nodes
> > > NAME                                 STATUS   ROLES    AGE    VERSION
> > > eu-de-2-master-0.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
> > > eu-de-2-master-1.cz-d8e4.cloud.lab   Ready    master   111m   v1.20.0+7d0a2b2
> > > eu-de-2-master-2.cz-d8e4.cloud.lab   Ready    master   112m   v1.20.0+7d0a2b2
> > > eu-de-2-worker-0.cz-d8e4.cloud.lab   Ready    worker   96m    v1.20.0+7d0a2b2
> > > eu-de-2-worker-1.cz-d8e4.cloud.lab   Ready    worker   95m    v1.20.0+7d0a2b2
> > > [root@cz-d8e4-eu-de-2-bastion-0 ~]#
> > > ```
> 
> Can cluster-id be empty ? For the case where user doesn't want the hexID

`cluster_id` cannot be empty because helpernode vars required it. So when `cluster_id` is not provided, random hex will be used.



> > I think this fits all the use cases. @bpradipt are we okay with it?
> > Suggestion:
> > `eu-de-2-master-0.cz-d8e4.cloud.lab` -> `master-0.cz-d8e4-eu-de-2.cloud.lab`
> > The region info can be added to the full cluster name and will be easy to implement.
> 
> For the above, @cs-zhang mentioned the following "For resource name, all resources belong to same cluster should be grouped together showing in UI. For Example 3: master0-bpradipt-dev-eu-de-2-idk4 that would not be possible, in this case the resources will be grouped by base node name master0. So I prefer bpradipt-dev-eu-de-2-idk4-master0, that will make all resources will be grouped together."

I am fine with the UI naming. This suggestion was specifically for `oc get nodes` for simplicity we can associate the region name with cluster_id.

**Comment by yussufsh** [comment]
UI: `cz-d8e4-eu-de-2-master-0`
Nodes: `master-0.cz-d8e4-eu-de-2.cloud.lab`

**Comment by cs-zhang** [comment]
> eu-de-2-master-0.cz-d8e4.cloud.lab -> master-0.cz-d8e4-eu-de-2.cloud.lab

Since the FQDN for the master-0 is `eu-de-2-master-0.cz-d8e4.cloud.lab`, the same as the node name and  the region info is not used in cluster_id. Change to `master-0.cz-d8e4-eu-de-2.cloud.lab` can cause confuse for user.  So for consistent with UI and FQDN,  following is better:
UI: cz-d8e4-eu-de-2-master-0
Nodes: eu-de-2-master-0.cz-d8e4.cloud.lab

If the `cluster_id` is defined as `ocp4`:
UI: ocp4-eu-de-2-master-0
Nodes: eu-de-2-master-0.ocp4.cloud.lab

**Comment by yussufsh** [comment]
> 
> 
> > eu-de-2-master-0.cz-d8e4.cloud.lab -> master-0.cz-d8e4-eu-de-2.cloud.lab
> 
> Since the FQDN for the master-0 is `eu-de-2-master-0.cz-d8e4.cloud.lab`, the same as the node name and the region info is not used in cluster_id. Change to `master-0.cz-d8e4-eu-de-2.cloud.lab` can cause confuse for user. So for consistent with UI and FQDN, following is better:
> UI: cz-d8e4-eu-de-2-master-0
> Nodes: eu-de-2-master-0.cz-d8e4.cloud.lab
> 
> If the `cluster_id` is defined as `ocp4`:
> UI: ocp4-eu-de-2-master-0
> Nodes: eu-de-2-master-0.ocp4.cloud.lab

I am fine with anything.. it was just a suggestion 😄  that would help while implementing the changes.

**Comment by yussufsh** [comment]
@cs-zhang Changes are good but I feel we can refactor the code to build the name_prefix and node_prefix (including the training `-`) in ocp.tf instead of each module. Also the cluster_domain with wildcard dns can be built in root module and passed to each module.
What do you think?

File: modules/4_nodes/nodes.tf [hunk 1]
```diff
@@ -186,9 +196,9 @@ resource "null_resource" "remove_worker" {
     when       = destroy
     on_failure = continue
     inline = [<<EOF
-oc adm cordon worker-${count.index}
-oc adm drain worker-${count.index} --force --delete-local-data --ignore-daemonsets --timeout=180s
-oc delete node worker-${count.index}
+oc adm cordon ${self.triggers.node_prefix}-worker-${count.index}.${self.triggers.cluster_id}.${self.triggers.cluster_domain}
```
Comment (yussufsh) [inline modules/4_nodes/nodes.tf:199]: Is it required to use the domain here? would not it get resolved without the domain part?

Comment (cs-zhang) [reply modules/4_nodes/nodes.tf:199]: It requires the node name, `eu-de-2-master-0.ocp4.cloud.lab` is node name, `eu-de-2-master-0` is not.

File: modules/4_nodes/nodes.tf [hunk 2]
```diff
@@ -17,6 +17,10 @@
 # limitations under the License.
 #
 ################################################################
+locals {
+  wildcard_dns   = ["nip.io", "xip.io", "sslip.io"]
```
Comment (yussufsh) [inline modules/4_nodes/nodes.tf:21]: Can we also move this to ocp.tf locals and pass it to nodes & install module?

File: modules/4_nodes/nodes.tf (hunk 1 above)
Comment (yussufsh) [reply modules/4_nodes/nodes.tf:196]: `node_prefix` will have a `-` at the end from ocp.tf right?

File: ocp.tf [hunk 3]
```diff
@@ -16,6 +16,7 @@ locals {
   # Generates cluster_id as combination of cluster_id_prefix + (random_id or user-defined cluster_id)
   cluster_id   = var.cluster_id == "" ? random_id.label[0].hex : (var.cluster_id_prefix == "" ? var.cluster_id : "${var.cluster_id_prefix}-${var.cluster_id}")
   name_prefix  = var.use_zone_info_for_names ? "${local.cluster_id}-${var.ibmcloud_zone}" : local.cluster_id
```
Comment (yussufsh) [inline ocp.tf:18]: Also add a `-` at the end in similar format to `node-prefix`

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: modules/4_nodes/nodes.tf (hunk 2 above)
Comment (cs-zhang) [reply modules/4_nodes/nodes.tf:21]: It is not referenced by any other modules, so no need to make it global.

Comment (yussufsh) [reply modules/4_nodes/nodes.tf:21]: You are passing it to install module right? So suggestion is to populate the domain in ocp.tf and pass it as domain_name to both modules via variables.

Comment (cs-zhang) [reply modules/4_nodes/nodes.tf:21]: No, just passed in domain_name created in node module.

File: ocp.tf (hunk 3 above)
Comment (cs-zhang) [reply ocp.tf:18]: I'll make it consistent with node_prefix.

**Review by yussufsh — APPROVED** [review]
/lgtm

**Review by ssinha20 — APPROVED** [review]
LGTM.

**Comment by ssinha20** [comment]
@yussufsh : Please let me know if you need any help with this issue/ we are blocked to close?

**Review by bpradipt — COMMENTED** [review]
/lgtm

**Review by bpradipt — APPROVED** [review]

**Pull request merged**